compliance_checker/cf/cf.py

```diff
@@ -677,6 +707,90 @@ def _split_standard_name(self, standard_name):
+                string describing the attribute.
+        """
+
+        if attr_type == 'S':
+            if not isinstance(attribute, basestring):
+                return [False,
+                        "{} must be a string".format('{}')]
+        else:
+            # if it's not a string, it should have a numpy dtype
+            underlying_dtype = getattr(attribute, 'dtype', None)
+
+            # TODO check for np.nan separately
+            if underlying_dtype is None:
+                return [False,
+                        "{} must be a numeric numpy datatype".format('{}')]
+
+            # both D and N should be some kind of numeric value
+            is_numeric = np.issubdtype(underlying_dtype, np.number)
+            if attr_type == 'N':
+                if not is_numeric:
+                    return [False,
+                            "{} must be numeric".format('{}')]
+            elif attr_type == 'D':
+                # TODO: handle edge case where variable is unset here
+                var_dtype = getattr(variable, 'dtype', None)
+                if (underlying_dtype != var_dtype):
+                    return [False,
+                            "{} must be numeric and must match {} dtype".format('{}', var_dtype)]
+            else:
+                # If we reached here, we fell off with an unrecognized type
+                return [False, "{} has unrecognized type '{}'".format('{}',
+                                                              attr_type)]
+        # pass if all other possible failure conditions have been evaluated
+        return [True, None]
+
+    def _handle_dtype_check(self, attribute, attr_name, attr_dict,
+                            variable=None):
+        """
+        Helper function for Appendix A checks.
+
+        :param attribute: The value of the attribute being checked
+        :param str attr_name: The name of the attribute being processed
+        :param dict attr_dict: The dict entry with type and attribute location
+                               information corresponding to this attribute
+        :param variable: if given, the variable whose type to check against
+        :rtype: tuple
+        :return: A two-tuple that contains pass/fail status as a boolean and
+                 a message string (or None if unset) as the second element.
+        """
+        attr_type = attr_dict['Type']
+        if variable is None and 'G' not in attr_dict['attr_loc']:
+            raise ValueError('Non-global attributes must be associated with a '
+                             ' variable')
+        attr_str = ('Global attribute {}'.format(attr_name)
+                    if 'G' in attr_dict['attr_loc'] and variable is None
+                    else "Attribute {} in variable {}".format(attr_name,
+                                                              variable.name))
+
+        # check the type
+        return_value = self._check_attr_type(attribute, attr_type, variable)
```

Comment:
I would just pass in `attr_str` to this function rather than reassigning the formatted value later.

Reply:
I understand where you're coming from here, but I wanted to make the helper function as atomic as possible so it could be used in other functions requiring attribute type checking. I think it's a really clear way to compose the function.